### PR TITLE
Add DLPack support for XPU backend by mapping to kDLOneAPI in DLPack

### PR DESCRIPTION
**Pull request by chengjunlu**
## Motivation
The DLPack device type kDLOneAPI stands for the Unified Shared Memory allocated on a oneAPI device. The corresponding Pytorch backend type is XPU. 
Support to export/import the Pytorch XPU tensor as a DLPack tensor of kDLOneAPI device.

## Solution
1. Update the DLPack protocol to v0.7.
2. Add the XPU hooks to map the Aten device and DLPack device with the address value and device information.

### CODE REVIEW

**Review by kit1980 — APPROVED** [review]

**Comment by kit1980** [comment]
@chengjunlu Heads up we will probably have to revert this again. We already use DLPack version 60 internally, and this PR adds DLPack version 70 causing a lot of issues.

**Comment by chengjunlu** [comment]
@kit1980 
When the internal dl v0.6 is available public?
I may work based on the v0.6 to v0.7.

**Comment by kit1980** [comment]
> @kit1980 When the internal dl v0.6 is available public? I may work based on the v0.6 to v0.7.

I believe it's just regular DLPack 0.6, just copied internally to Meta. If you can create a new version of PR without updating DLPack to 0.7, it will be easier to merge. But please ping me on the PR.

**Comment by chengjunlu** [comment]
> > @kit1980 When the internal dl v0.6 is available public? I may work based on the v0.6 to v0.7.
> 
> I believe it's just regular DLPack 0.6, just copied internally to Meta. If you can create a new version of PR without updating DLPack to 0.7, it will be easier to merge. But please ping me on the PR.

I have to use the DLPack 0.7 because the kDLoneAPI is defined from that.

What do you think should I suppose to do? Wait the internal DLPack 0.6 changes to be available in public?

**Comment by kit1980** [comment]
> I have to use the DLPack 0.7 because the kDLoneAPI is defined from that.
> 
> What do you think should I suppose to do? Wait the internal DLPack 0.6 changes to be available in public?

There is no internal DLPack 0.6 changes. There is code outside of PyTorch internally that includes DLPack 0.6.
And because Meta has monorepo, we have a situation that with this PR both DLPack 0.6 and DLPack 0.7 needs to be included - which is a problem.
Also there is code internally that defines `kDLWebGPU`, for example. But that is already defined in DLPack - so again a problem.

I think the best way forward is to update internal usage to DLPack 0.7 before merging this PR.
I'll work on this, but not sure about how much time this will take.

**Comment by kit1980** [comment]
@chengjunlu FYI, we're working on changes to allow to land this PR smoothly. I'll let you know when we're ready.

**Comment by chengjunlu** [comment]
> @chengjunlu FYI, we're working on changes to allow to land this PR smoothly. I'll let you know when we're ready.

:) Thanks for the information.

**Comment by chengjunlu** [comment]
> @chengjunlu FYI, we're working on changes to allow to land this PR smoothly. I'll let you know when we're ready.

Hi @kit1980,
Any new information about this?

**Comment by kit1980** [comment]
> > @chengjunlu FYI, we're working on changes to allow to land this PR smoothly. I'll let you know when we're ready.
> 
> Hi @kit1980, Any new information about this?

Not yet. We need to change how DLPack is used internally in Meta. I have this task near the top of my priority list.

**Comment by chengjunlu** [comment]
> > > @chengjunlu FYI, we're working on changes to allow to land this PR smoothly. I'll let you know when we're ready.
> > 
> > 
> > Hi @kit1980, Any new information about this?
> 
> Not yet. We need to change how DLPack is used internally in Meta. I have this task near the top of my priority list.

It is glad to hear that. Looking forward the changes. Thank for the help.

**Comment by chengjunlu** [comment]
> > > @chengjunlu FYI, we're working on changes to allow to land this PR smoothly. I'll let you know when we're ready.
> > 
> > 
> > Hi @kit1980, Any new information about this?
> 
> Not yet. We need to change how DLPack is used internally in Meta. I have this task near the top of my priority list.

Hi @kit1980, 
A  kindly remaind. 
As the torch 1.13 is going to be branched out recently, we'd like to add the support the DLPack for XPU backend.
May I know the status about the latest DLPack in pytorch master?

**Comment by kit1980** [comment]
> Hi @kit1980, A kindly remaind. As the torch 1.13 is going to be branched out recently, we'd like to add the support the DLPack for XPU backend. May I know the status about the latest DLPack in pytorch master?

The branch cut is end of this week. Realistically, I don't think it can make into 1.13. It requires changes to Meta-internal code outside of PyTorch. I need to finish some other tasks and then I'll continue with this.

Can you maybe try without updating DLPack? Maybe define the kDLoneAPI outside of the DLPack?

**Comment by gujinghui** [comment]
@kit1980 
PyTorch 1.13 is coming. We know this change will not in 1.13.
May I know whether we can make it happen in next release, say, PyTorch 2.0?
Thanks a lot.

**Comment by kit1980** [comment]
@gujinghui I'll try to unblock this before the next major release.

**Comment by gujinghui** [comment]
> @gujinghui I'll try to unblock this before the next major release.

Great news. Many thanks.

**Comment by guangyey** [comment]
Hi @kit1980, I reopen this PR in #94968.  Please help review and any comments are welcome.

**Pull request closed without merging**
